### ISSUE TITLE
Multiple issues for v0.7.4dev

### ISSUE THREAD

**Issue by albertelonjobs** [cutright/DVH-Analytics#50]
Critical Issues:
1. In ROI Map Editor, when "DEFAULT" Physician is selected, a new institutional ROI is added but it does not show up in the list.
2. In import DICOM window, the "ROI Manager" button doesn't work. I got the error message: "AttributeError: 'DatabaseROIs' object has no attribute 'physicians'". I guess it might be caused by physician name is missing in the DICOM plan file.
3. In import DICOM window, once a ROI label has been assigned to a RT structure, it disappears from the dropdown list. This is a problem for PTV and CTV. Because the same label could be assigned to multiple structures.
4. A physician's ROI table needs to have a copy of every labels in the institutional ROI table, even though the physician has no variations for some structure labels. This requirement is very inconvenient for physician tables that have already been created and are missing labels compared to the new institutional table. In other words, if the institutional table is updated, all physician tables have to be individually updated, which is very annoying.

Minor Issues:
1. In ROI Map Editor, after a Physician ROI is added, the dropdown box for the Physician was reset to DEFAULT, which is annoying.
2. In ROI Map Editor, after "Save and Update Database" button is clicked, the ROI Map window should not be automatically closed.
3. In ROI Map Editor, allow a variation label to be edited.
4. The multiple consecutive spaces/underscores in a structure label should be pre-processed and be treated as a single space/underscore. For example, "CTV Prostate" and "CTV  Prostate" should be considered as the same label variation.

**Comment by cutright** [comment]
Thanks for the detailed info and the beta testing!

**Critical**:
 - [x] 1. Looking into it.  I haven't focused on DEFAULT since I imagine most users will want to store a physician in the database.  I'm considering hiding DEFAULT physician from the GUI.
 - [X] 2. Fixed
 - [X] 3. Addressed with right-clicking ROI (note added in GUI). Should add an alert if importing multiple instances of a physician ROI, if not a target.
 - [ ] 4. I'm not sure I understand.  If I force every physician to adopt the Institutional names, is there a purpose to having a physician map?  This could also run into conflicts limiting what a physician can have in his/her map.

**Minor**:
 - [x] 1. Should be an easy fix.
 - [X] 2. Fixed
 - [ ] 3. I think I'll actually pass on this.  It would only save one mouse click.  Since there are no children objects of a variation, it's the same as just deleting it then adding what you want.
 - [x] 4. I think this is an easy fix given the coding design (fixed?)

And finally... I can assure you that building an ROI map is by far the most frustrating part of the process... but things work out very smoothly once in place.

We're also working on some technology to make some of this superfluous.
https://uchicago.technologypublisher.com/technology/36434

**Comment by cutright** [comment]
I should also point out that space and underscore alterations are not stored in the ROI Map explicitly... so you can add A_Aorta or a__aorta... but you can not add both.  I did this so that your ROI map looks the way you want (i.e., TG-263 names were hard to look at in all lower case).

However, when DVHA looks up the Physician ROI of an ROI in a new plan, it will apply the clean_name function below and find the Physician ROI with an equivalent clean_name, then return the original Physician ROI.

https://github.com/cutright/DVH-Analytics/blob/c89f6faf9f097d9ba41b667052b01da621e945f3/dvha/tools/roi_name_manager.py#L732-L739

**Comment by albertelonjobs** [comment]
Critical issue #1:  A physician may want to define a structure that doesn't have a DEFAULT label. For example, if a structure that combines the left and right femur heads, there is no structure label for that in the institutional labels. So the new label can only be linked to "uncatergorized" in the institutional ROI. Is this the intention of the system design?
Critical issue #3: What about if a composite plan that include both the initial and the boost plan? It has an initial PTV and a boost PTV. How to handle that?
Critical issue #4: This is the problem I have. I'm trying to import a plan with no physician info in the plan but I know which physician it belongs to. In the Import DICOM window, if I click the Apply button with "DEFAULT" physician selected, the physician specific labels are not applied. But If I choose the physician label, the institutional labels which are not in the physician label list are not recognized. I cannot have it both ways unless every institutional labels are in this physician's label list.

**Comment by albertelonjobs** [comment]
One more critical issue to report:
5. A variation of PTV is created in a physician's label and linked to the institutional label "PTV". However in Import DICOM window, the ROI Type for the a physician's PTV label is "ORGAN" instead of "PTV", which results an automatic no PTV found warning.

**Comment by cutright** [comment]
Critical #3: DVHA will find all structures with an ROI Type of PTV, create a composite structure for the purpose of distance and overlap calculations.  It also assigns ROI Types of PTV1, PTV2, PTV3, etc. in order of increasing V95%.  If you need something more specific you can edit your ROI Map as you like, just be sure each PTV is assigned an ROI Type of PTV on import.  If you had assigned the structure as an ROI Type of PTV in your TPS, DVHA will default to that.  Otherwise it will auto assign based on this function:

https://github.com/cutright/DVH-Analytics/blob/270ca5f657f7e694ee59d3c0cc28c10ad17c35a9/dvha/db/dicom_parser.py#L1628-L1649

Basically looks for PTV, PTV#, PTV #... case insensitive.  And you can always override it.

The ROI Map does not actually do anything about ROI Type.  

The DVHA manual can be found on the releases page: https://github.com/cutright/DVH-Analytics/releases,  
currently the latest one is 
https://github.com/cutright/DVH-Analytics/releases/download/v0.7.3/DVH.Analytics.Manual.pdf

**Comment by cutright** [comment]
Critical #4: You can only apply one "physician" at a time, which is intentional.  I'm just not sure there's an easy way to address this?  It sounds like your physician specific map is incomplete to me.

The problem with auto adding new Institutional ROIs to physician maps is what if you add an institutional ROI that is very similar to something another physician is already using, then that physician has two names for the same organ.  An example of this issue is in the BBM map if you click "Branched ROIs" in the top left on the ROI Map screen.  I believe the DVHA GUI no longer allows you to create a map like this.

![image](https://user-images.githubusercontent.com/4778878/76454815-34826000-63a3-11ea-9e67-8d1c390ec9ef.png)


I think the best way to approach this is to develop each physician map individually.  But maybe build one good one first, then copy it and edit it.  I wish it were easier... but data curation is hard.

Perhaps what I could do is present a pop-up window with a list of unused Institutional ROIs for a physician?  Let the user add selected ones (with CTRL or SHIFT clicks).  I'll need to think about some GUI elements go from there.

I also recommend selecting an institutional ROI every time you add a physician ROI.  Perhaps I could suggest it in a similar way I do on the add variation on the import screen.

**Comment by albertelonjobs** [comment]
Critical #3: I understand how the automatic assignment works. But suppose I'm trying to import a new variation of PTV which has not been defined in the physician's label list. I would be unable to manually choose the correct label from the dropdown list in the Physician's ROI label because the label "ptv" has already been taken by another PTV. So the label for the new PTV has to be "uncategorized". I understand I can still choose "PTV" as its ROI Type. Do you think this is no big deal?

**Comment by cutright** [comment]
I think I understand now.  I should remove the duplicate Physician/Institutional ROI import restriction for targets.  Would that address the issue?

**Comment by cutright** [comment]
Oh... I see the real issue.

The drop down for Physician's ROI Label in the bottom right only displays Physicain ROIs not yet used.

However, you can right-click any name, and add it as a variation to any Physicians ROI.

There is some over-looked details here, but you can in fact do what you like I think?

**Comment by albertelonjobs** [comment]
Yes. I didn't know the trick of right-click. Thank you very much!

**Comment by cutright** [comment]
Sorry for the confusion...

There are pros and cons to both sets of logic (right-click vs drop-down).  Obviously they need some tweaks.  The drop down is nice because you'll quickly see which physician ROIs you are missing.  I'll have to think about how to tidy this up.

**Comment by albertelonjobs** [comment]
Would you be able to look into the critical issue #5 I reported a while ago?

Critical #5: A variation of PTV is created in a physician's label and linked to the institutional label "PTV". However in Import DICOM window, the ROI Type for the a physician's PTV label is "ORGAN" instead of "PTV", which results an automatic no PTV found warning.

**Comment by cutright** [comment]
Can you give me a specific example?

I *think* DVHA only auto-detects based on the name or type in the DICOM dataset.  I could have it also check its associated physician and institutional roi.

**Comment by albertelonjobs** [comment]
Here are the files for the physician's label table and DICOM files. Try to import the DICOM file and apply physician RA and check the structures "PTV46 PX_SV_LN" and "PTV78 PX".

[test.zip](https://github.com/cutright/DVH-Analytics/files/4320672/test.zip)

**Comment by cutright** [comment]
Thanks, that helps. I think this amounts to 1 issue and 1 feature request
**Issue**:  These specific structure names should be auto detected as PTV regardless of your ROI Map.  Will modify the function for that.
**Feature Request**: The ROI Map should contain an ROI Type property for Physician ROIs and Institutional ROIs that the DICOM importer can look up. (More work to do this, but a good idea).

**Comment by albertelonjobs** [comment]
Great! I'm looking forward to it.

**Comment by cutright** [comment]
So it turns out the code already assigns ROI type if the physician ROI is gtv, ctv, itv, or ptv (case insensitive) as shown in the autodetect function previously posted.  However, this function was only running on initialization of the import screen.

Now this function runs every time you change the physician (and apply it).  I think it will behave in the way you want now?  And I think this makes assigning ROI Types in the ROI map a lower priority now?  That could be a lot of work to implement.  Alternatively, I could also apply some string logic to the physician ROI so it doesn't have to be one of the target types exactly.

Thanks again for the feedback.  Being the only programmer on this project, I don't get a chance to bug test as much as I'd like.

**Comment by cutright** [comment]
Now your institutional ROIs are saved to a physician_DEFAULT.roi.  This is not done in an ideal way, but due to the way I calculate ROI Map changes, the DEFAULT physician roi file needs to be in the same format.

Also, now when you delete an Institutional ROI from DEFAULT, it will propagate to any physician ROIs that used it, setting their institutional ROIs to uncategorized.

You'll notice that saving the ROI Map on the import screen is much faster, that's because it currently does not look to see if your changes affect what is already in your database and update them (unlike the ROI Map screen does... supposedly... needs to be validated and it's progress dialog needs updating).

And finally, there are some sorting issues that cause the ROI Map visual to behave a little strangely depending on how the map was edited.  This appears to be purely cosmetic.

**Comment by cutright** [comment]
Whenever you have a chance, let me know what you think of the latest version of v0.7.4dev?

I think the only thing I haven't really addressed is Critical 4 (which isn't trivial). If all bugs are fixed, I'd like to go ahead and push to master for a release since there are many substantial improvements since v0.7.3.

For others using PyPI, the state of v0.7.4dev branch has been pushed as a release candidate to PyPI as 0.7.4rc2
https://pypi.org/project/dvha/0.7.4rc2/

**Comment by albertelonjobs** [comment]
Thank you very much for your efforts! I'd like to try it today. But I couldn't find the branch v0.7.4rc2 in my GitHub desktop app. How can I download the source code? I have trouble to install it using pip.

**Comment by cutright** [comment]
For GitHub, just update v0.7.4dev, no new branch

**Comment by albertelonjobs** [comment]
I have imported 8 plans. But I could not get a DVH shown in the window. I have studied the manual but could not figure out the reason. Please help!
![Capture_DVH](https://user-images.githubusercontent.com/61991041/76556152-619c4480-646f-11ea-9fee-1d6b4fea6864.JPG)

**Comment by albertelonjobs** [comment]
By the way, I couldn't see the ROI map plot that you showed me above yesterday either.

**Comment by albertelonjobs** [comment]
In your future release, it might be a good idea to provide a few examples that a user can try to import and test the functionalities of the system.

**Comment by cutright** [comment]
This sounds like an issue unrelated to the new code.

1. Are you on the same computer as yesterday?
2. Can you try switching the master branch and perform a query?
3. Can you see which version of Bokeh you are using?
4. Navigate to your user folder -> /Apps/dvh_analytics/data/temp  ... are there any html files? Do they open in your browser?

Basically, all plots are generated with the bokeh library, which generates an html file.  DVHA then uses an IE emulator to view it.  I haven't seen your issue on Windows 7 or 10 in a very long time.  It could be that DVHA was not successfully able to change your windows registry to allow IE Emulation in v11.

**Comment by cutright** [comment]
Also, is anything printed in your terminal?

**Comment by albertelonjobs** [comment]
I downgraded bokeh from 2.0.0 to 1.4.0 and it works! Thank you very much!

**Comment by cutright** [comment]
Awesome!

Good to know.  Bokeh 2.0.0 appears to work fine on macOS, I guess I haven't tried on Windows yet (Bokeh 2.0.0 just came out in the past week or so).  I will verify on my end, and then set a requirement of Bokeh 1.4.0 for DVHA.

**Comment by cutright** [comment]
I also have an issue with bokeh==2.0.0 on Windows 7.  I'm actually unfamiliar with its new features, so DVHA doesn't need them.  I'll change the requirements.txt to bokeh==1.4.0.

**Comment by cutright** [comment]
Sounds like you've gotten through the ROI Mapping process, which is what 0.7.4 is really aiming to address.  I will likely push this to master tomorrow and create new executables, unless you come across something major?

**Comment by albertelonjobs** [comment]
The ROI mapping process runs good now except for a couple of minor issues.

1. Sometimes it fails to recognize the ROI type of a PTV  even though the ROI label is recognized by the system. The study is shown by the orange warning color. But once I clicked the PTV structure in the "ROI Mapping for Selected Study" view box, it is immediately recognized. I somehow have trouble to reproduce the problem. 
2. Studies panel is not auto populated when the Import DICOM window is opened, even though a correct import directory is shown in the text box. Please implement auto population or add a "Search" button.

**Comment by cutright** [comment]
1. I think I understand, but I'll need to be able to reproduce it to know what code needs to be changed.
Are you using Eclipse by chance?  I wonder if it has to do with having one structure set across multiple RT Plan files?

2. I already have functionality coded to auto-parse on import screen launch, but now that DICOM parsing is queued/threaded to avoid memory management issues with Windows, I need to implement proper cancellation.  Unfortunately, closing the "Reading DICOM Headers" or "Parsing File Sets" windows doesn't kill the threads and will throw errors.  Once I implement thread cancellation for these two, I'll add the auto-parse on import screen launch back.

I'm going to table these two for another release since there's a ton of fixes ready to share.

Thanks again... please keep me posted on what you find!

**Comment by cutright** [comment]
Just an FYI, since I've merged 0.7.4dev into master and I've deleted the 0.7.4dev branch, so you'll get an error if you try to update the dev... just switch over to the master branch.

Moving forward, it's probably best for you to stick with the master branch since a dev branch may not always be stable/usable... unless there's something new/interesting to test out early.

**Comment by albertelonjobs** [comment]
Thank you very much!
